alerting/alerter-war/src/main/java/org/hawkular/metrics/alerter/ExpressionManager.java

```diff
@@ -81,31 +94,91 @@
     @Inject
     private AlertsService alerts;
 
+    /**
+     * Access to the manager of the caches used for the partition services.  This is used to inspect cluster
+     * topology and ensure we only run the external alerter on one node (the coordinator). Otherwise
+     * each node would execute the same work, resulting in duplicate alerts.
+     */
+    @Resource(lookup = "java:jboss/infinispan/container/hawkular-alerts")
+    private EmbeddedCacheManager cacheManager;
+
     @PostConstruct
     public void init() {
-        log.debug("Initializing Hawkular Alerts-Metrics Manager...");
+        // Cache manager has an active transport (i.e. jgroups) when is configured on distributed mode
+        distributed = cacheManager.getTransport() != null;
+
+        if (distributed) {
+            topologyChangeListener = new TopologyChangeListener();
+            cacheManager.addListener(topologyChangeListener);
+        }
+
+        processTopologyChange();
+    }
+
+    // When a node is joining/leaving the cluster the coordinator node could change
+    @Listener
+    public class TopologyChangeListener {
+        @ViewChanged
+        public void onTopologyChange(ViewChangedEvent cacheEvent) {
+            processTopologyChange();
+        }
+    }
+
+    private void processTopologyChange() {
+        boolean currentCoordinator = coordinator;
+        coordinator = distributed ? cacheManager.isCoordinator() : true;
+
+        if (coordinator && !currentCoordinator) {
+            start();
```

Comment:
Logic looks good.
The only think I wonder is about the speed.
I mean, jgroups tends to move the coordinator, so on startups a real possibility is that the new node becomes coordinator.
So, you can count that stop() and start() can be launched separately.
My only doubt if those process are safe or if they need some kind of sync.
I.e. a start() can not happen until a stop() in other node is fully finished.
If that is ok, the code is correct.
If not, perhaps some locking is necessary.